### PR TITLE
#760 Allow providers to impersonate users and record the provider in the history 

### PR DESCRIPTION
**Pull request by doolse**
#760

### CODE REVIEW

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/core/oauthserver/OAuthServerAccess.scala [hunk 1]
```diff
@@ -155,32 +156,44 @@ object OAuthServerAccess {
       .authenticateAsUser(username, LegacyGuice.userService.getWebAuthenticationDetails(request))
   }
 
+  def lookupCloudToken(tokenData: String,
+                       request: HttpServletRequest): DB[Option[ModifiableUserState]] = {
+    val (actualToken, impersonateId) = tokenData.indexOf(';') match {
```
[the remark this answers is not in the document]
Comment (doolse) [reply Source/Plugins/Core/com.equella.core/scalasrc/com/tle/core/oauthserver/OAuthServerAccess.scala]: The choice of separater character is arbitrary, as long as it doesn't appear in a UUID string

File: Source/Plugins/Core/com.equella.serverbase/scalasrc/com/tle/core/db/dao/EntityDB.scala [hunk 2]
```diff
@@ -71,11 +71,14 @@ object EntityDB {
     }
   }
 
+  def readOneStream[A](uuid: UUID)(implicit ee: EntityDBExt[A]): Stream[DB, A] =
```
[the remark this answers is not in the document]
Comment (doolse) [reply Source/Plugins/Core/com.equella.serverbase/scalasrc/com/tle/core/db/dao/EntityDB.scala]: In so much as it's being called from the impersonation code, yes :)

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/core/oauthserver/OAuthServerAccess.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (SammyIsConfused) [reply Source/Plugins/Core/com.equella.core/scalasrc/com/tle/core/oauthserver/OAuthServerAccess.scala]: But on line 152 on HistoryEvent.java is specifies a colon, and here a semicolon?

File: Source/Plugins/Core/com.equella.serverbase/scalasrc/com/tle/core/db/dao/EntityDB.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (SammyIsConfused) [reply Source/Plugins/Core/com.equella.serverbase/scalasrc/com/tle/core/db/dao/EntityDB.scala]: Ah yeah I see the function call now. All good, sorry!

**Comment by SammyIsConfused** [comment]
hey @doolse, I'm pretty sure the autotest failures on codebuild are directly caused by this branch, seems to be an NPE related to user accounts.

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/core/oauthserver/OAuthServerAccess.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (doolse) [reply Source/Plugins/Core/com.equella.core/scalasrc/com/tle/core/oauthserver/OAuthServerAccess.scala]: That's for storing the impersonation string inside the history event user column, it's separate from the oauth token header.

**Comment by doolse** [comment]
> hey @doolse, I'm pretty sure the autotest failures on codebuild are directly caused by this branch, seems to be an NPE related to user accounts.

Having a look now..

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/core/oauthserver/OAuthServerAccess.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (SammyIsConfused) [reply Source/Plugins/Core/com.equella.core/scalasrc/com/tle/core/oauthserver/OAuthServerAccess.scala]: Gotcha, had my wires crossed there. Resolved.

**Pull request merged**